astacus/coordinator/plugins/flink/steps.py

```diff
@@ -0,0 +1,134 @@
+                if result:
+                    results.update(result)
+            if change_watch.has_changed:
+                raise TransientException("Concurrent modification during access entities retrieval")
+        return results
+
+    async def _get_result(self, zk_path, change_watch, connection):
+        result = {}
+        if await connection.exists(zk_path):
+            result[zk_path] = await self._get_data_recursively(
+                zk_prefix=zk_path, current_node=zk_path, connection=connection, watch=change_watch
+            )
+        return result
+
+    async def _get_data_recursively(
+        self, zk_prefix: str, current_node: str, connection: ZooKeeperConnection, watch: ChangeWatch
+    ):
+        result = {}
+        if await connection.exists(zk_prefix):
+            children = await connection.get_children(current_node, watch=watch)
+            for child in children:
+                value = await connection.get(f"{current_node}/{child}", watch=watch)
+                if value:
+                    result[child] = value.decode("UTF-8")
+                else:
+                    result[child] = await self._get_data_recursively(zk_prefix, f"{current_node}/{child}", connection, watch)
+        return result
+
+
+@dataclasses.dataclass
+class CreateFlinkManifestStep(Step[FlinkManifest]):
+    """
+    Collects data from previous steps into a `FlinkManifest`.
+    """
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> FlinkManifest:
+        return FlinkManifest(data=context.get_result(RetrieveDataStep))
+
+
+@dataclasses.dataclass
+class FlinkManifestStep(Step[FlinkManifest]):
+    """
+    Extracts the Flink plugin manifest from the main backup manifest.
+    """
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> FlinkManifest:
+        backup_manifest = context.get_result(BackupManifestStep)
+        return FlinkManifest.parse_obj(backup_manifest.plugin_data)
+
+
+@dataclasses.dataclass
+class UploadFlinkManifestStep(Step[None]):
+    """
+    Store the backup manifest in the object storage.
+
+    """
+    json_storage: AsyncJsonStorage
+    plugin: ipc.Plugin
+    plugin_manifest_step: Optional[Type[Step[AstacusModel]]] = None
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
+        if self.plugin_manifest_step is None:
+            plugin_data = {}
+        else:
+            plugin_manifest = context.get_result(self.plugin_manifest_step)
+            plugin_data = plugin_manifest.dict()
+        manifest = ipc.BackupManifest(
+            attempt=context.attempt,
+            start=context.attempt_start,
+            snapshot_results=[],
+            upload_results=[],
+            plugin=self.plugin,
+            plugin_data=plugin_data,
+        )
+        logger.debug("Storing backup manifest %s", context.backup_name)
+        await self.json_storage.upload_json(context.backup_name, manifest)
```

Comment:
I think it would be better to avoid the duplication of code and instead make `UploadManifestStep` a little bit more configurable so that you can disable the snapshot and update steps:
```python
@dataclasses.dataclass
class UploadManifestStep(Step[None]):
    """
    Store the backup manifest in the object storage.

    The backup manifest contains the snapshot from the `SnapshotStep` as well as the
    statistics collected by the `UploadBlocksStep` and the plugin manifest.
    """
    json_storage: AsyncJsonStorage
    plugin: ipc.Plugin
    plugin_manifest_step: Optional[Type[Step[AstacusModel]]] = None
    snapshot_step: Optional[Type[Step[List[ipc.SnapshotResult]]]] = SnapshotStep
    upload_step: Optional[Type[Step[List[ipc.SnapshotUploadResult]]]] = UploadBlocksStep

    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
        if self.plugin_manifest_step is None:
            plugin_data = {}
        else:
            plugin_manifest = context.get_result(self.plugin_manifest_step)
            plugin_data = plugin_manifest.dict()
        manifest = ipc.BackupManifest(
            attempt=context.attempt,
            start=context.attempt_start,
            snapshot_results=context.get_result(self.snapshot_step) if self.snapshot_step else [],
            upload_results=context.get_result(self.upload_step) if self.upload_step else [],
            plugin=self.plugin,
            plugin_data=plugin_data,
        )
        logger.debug("Storing backup manifest %s", context.backup_name)
        await self.json_storage.upload_json(context.backup_name, manifest)
```